code.js

```diff
@@ -1,3 +1,53 @@
 function tsp_hk(distance_matrix) {
-    return -1;
+    tsp_mems = [];
+    if(distance_matrix.length < 1) {
+        return 0;
+    } else {
+    for(let i = 0; i < distance_matrix.length; i++){
+        temp = tsp_HeldKarp(distance_matrix, i)
+    }
+    return temp
+    }
+}
+
+function tsp_HeldKarp(distance_matrix, start) {
+    if(tsp_mems[JSON.stringify(distance_matrix) + start] === undefined) {
```

Comment:
What if the distance matrix is in a different order? Shouldn't that go to the same cache entry?

Reply:
So I fixed the issue with the matrix shrink function. As for the issue with the different orders of distance matrices, I'm now using the nodesLeft array with the start to memoize. Do I need to effectively sort the start into the nodes left to make sure they do not do the work for repetitive information.

Reply:
You could do that, or convince me that it'll only ever be a certain order :)